### PR TITLE
virtualbox: discover extension packs at runtime

### PR DESCRIPTION
**Pull request by flokli**
###### Motivation for this change

Currently, `virtualbox` with extension pack and `virtualbox` without extension pack need to be build separately, because it's just an argument in the same derivation leaving to a different `installPhase`.

As the extension pack is unfree (and unfree isn't built by hydra), this means users with extension pack enabled need to compile `virtualbox` by themselves quite often.

This changes virtualbox to load extension packs from a environment variable `VBOX_EXTPACK_DIR`, so we can build virtualbox once by hydra, and let users of the extension pack just use the wrapper in front of it.

See https://github.com/NixOS/nixpkgs/issues/34796#issuecomment-406095992 for the original idea.

However, this might not work with enabled hardening, as the environment variable will be ignored. To fix it there, we might want to hardcode `vboxExtpackDir` to some imperative location changed during configuration switch, or replace the wrapper with something binary-patching the original virtualbox.

This draft PR is also to get the discussion started.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers 
cc @ambrop72 @xeji @cdepillabout

### CODE REVIEW

File: pkgs/applications/virtualization/virtualbox/default.nix [hunk 1]
```diff
@@ -7,7 +7,7 @@
 , javaBindings ? true, jdk ? null # Almost doesn't affect closure size
 , pythonBindings ? false, python3 ? null
 , extensionPack ? null, fakeroot ? null
-, pulseSupport ? config.pulseaudio or stdenv.isLinux, libpulseaudio ? null
+, pulseSupport ? false, libpulseaudio ? null
```
Comment (Mic92) [inline pkgs/applications/virtualization/virtualbox/default.nix]: Was this change intentionally?

Comment (flokli) [reply pkgs/applications/virtualization/virtualbox/default.nix]: Right - if `config.pulseaudio` is false, we'd previously build a non-pulseaudio variant.

What about including `libpulseaudio` in `buildInputs` if `headless` is `true`? I doubt we save that much by building non-pulseaudio, but graphical virtualbox.

Comment (flokli) [reply pkgs/applications/virtualization/virtualbox/default.nix]: updated.

**Comment by worldofpeace** [comment]
TBH, what's really gained in hardening if it's from an imperative location? Isn't that just as poor as the environment variable?

**Comment by ambrop72** [comment]
I suggest to solve this using a wrapper in one of the following ways:
- Make a trivial wrapper that sets VBOX_EXTPACK_DIR and exec's the real binary. This wrapper is itself wrapped by the setuid wrapper. It must be written in C, not bash, as bash refuses to run setuid.
- Extend the capabilities of the setuid wrapper system to allow the wrapper to set environment variables.

**Comment by ambrop72** [comment]
@worldofpeace It can still be secure if we are careful that both with and without the extension pack, the trusted wrapper always sets `VBOX_EXTPACK_DIR`. This way no matter what `VBOX_EXTPACK_DIR` is when calling the setuid program, it will be overwritten with the hardcoded value by wrapper.

**Comment by worldofpeace** [comment]
Those are great ideas @ambrop72 :sparkles: 
Now if only make wrapper was a C wrapper too... :smile:

File: pkgs/applications/virtualization/virtualbox/extpack.nix [hunk 2]
```diff
@@ -1,20 +1,28 @@
...
-fetchurl rec {
-  name = "Oracle_VM_VirtualBox_Extension_Pack-${version}.vbox-extpack";
-  url = "https://download.virtualbox.org/virtualbox/${version}/${name}";
-  sha256 =
-    # Manually sha256sum the extensionPack file, must be hex!
-    # Thus do not use `nix-prefetch-url` but instead plain old `sha256sum`.
-    # Checksums can also be found at https://www.virtualbox.org/download/hashes/${version}/SHA256SUMS
-    let value = "27a0956940654b0accf4d79692078bd496d9f062e4ed3da69e5421cba8d1e444";
-    in assert (builtins.stringLength value) == 64; value;
+stdenv.mkDerivation rec {
```
[the remark this answers is not in the document]
Comment (flokli) [reply pkgs/applications/virtualization/virtualbox/extpack.nix]: IIRC, Virtualbox wants to have that `$pname` in `VBOX_EXTPACK_DIR`. We could do all the extracting in the wrapper, but…

I also like it being a proper derivation, with all the ways to modify, fixup and patch it.

On top of that, we could start supporting multiple virtualbox extension packs.
The `virtualbox` derivation currently also produces a `$out/libexec/virtualbox/ExtensionPacks/Oracle_VBoxDTrace_Extension_Pack` - I'm not sure if anybody is using that, but moving it to a separate output and adding it to the list of exposed extpacks sounds useful.

**Comment by SRGOM** [comment]
I feel stupid commenting when there are much more knowledgeable people here but I have a question about: 

> we might want to hardcode vboxExtpackDir to some imperative location changed during configuration switch

Isn't the location of guest additions always known (even if guest additions aren't compiled)? The `.drv` derivation itself is something that hydra wouldn't mind making and that's what decides the store location ..., so wouldn't it work to compute what path to hardcode?

I should probably shut up ...

**Comment by flokli** [comment]
closing due to #74503. I'll keep the branch around if anybody wants to pick this up.

**Comment by kini** [comment]
FWIW, I've attempted to rebase this PR branch on the latest master: https://github.com/NixOS/nixpkgs/compare/master...kini:nixpkgs:virtualbox-extpack

**Pull request closed without merging**